flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/source/IcebergSource.java

```diff
@@ -357,13 +358,10 @@ public IcebergSource<T> build() {
       if (readerFunction == null) {
         RowDataReaderFunction rowDataReaderFunction =
             new RowDataReaderFunction(
+                (SerializableTable) SerializableTable.copyOf(table),
                 flinkConfig,
-                table.schema(),
                 context.project(),
-                context.nameMapping(),
```

Comment:
this essentially deprecates `ScanContext#nameMapping` and only retrieve from table property. Technically, this breaks backward compatibility. is there any use case where Flink job needs to set a different name mapping than table property?

Reply:
I remember that when we were implementing historical queries in Hive (AS OF VERSION, AS OF TIMESTAMP) then we found that the namemapping is bound to the table, and not bound to the version. So if some specific schema evolution happens (migrate a table, rename a column, add back an old column - or something like this - sadly I do not remember the specifics) then we were not able to restore the original mapping from the current one, and we do not able to query the data.
Being able to provide a namemapping could help here.

Arguably, this is a rare case, and the correct fix would be a spec change, still I thought it worth to mention.

Reply:
Also, it was long ago, so we might want to double check the current situation before acting on this 😄

Reply:
IIRC, This one (https://github.com/apache/iceberg/pull/2275) does the job. Now we can use schema Id in the snapshot to track which schema it writes. Anyway, let me revert this back since it should be in another PR even if we want to remove this.